reframe/core/schedulers/slurm.py

```diff
@@ -692,3 +724,316 @@ def _extract_attribute(self, attr_name, node_descr, sep=None):
+                 f'{self._system_name}')
+
+        self._push_artefacts(job)
+
+        intervals = itertools.cycle([1, 2, 3])
+        while True:
+            try:
+                # Make request for submission
+                submission_result = self.client.submit(
+                    self._system_name,
+                    os.path.join(job._remotedir, job.script_filename),
+                    local_file=False
+                )
+                break
+            except fc.FirecrestException as e:
+                stderr = e.responses[-1].json().get('error', '')
+                error_match = re.search(
+                    rf'({"|".join(self._resubmit_on_errors)})', stderr
+                )
+                if not self._resubmit_on_errors or not error_match:
+                    raise
+
+                t = next(intervals)
+                self.log(
+                    f'encountered a job submission error: '
+                    f'{error_match.group(1)}: will resubmit after {t}s'
+                )
+                time.sleep(t)
+
+        job._jobid = str(submission_result['jobid'])
+        job._submit_time = time.time()
+
+    def allnodes(self):
+        raise NotImplementedError('firecrest slurm backend does not support '
+                                  'node listing')
+
+    def filternodes(self, job, nodes):
+        raise NotImplementedError(
+            'firecrest slurm backend does not support node filtering'
+        )
+
+    def poll(self, *jobs):
+        '''Update the status of the jobs.'''
+
+        if jobs:
+            # Filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        poll_results = self.client.poll(
+            self._system_name, [job.jobid for job in jobs]
+        )
+        job_info = {}
+        for r in poll_results:
+            # Take into account both job arrays and heterogeneous jobs
+            jobid = re.split(r'_|\+', r['jobid'])[0]
+            job_info.setdefault(jobid, []).append(r)
+
+        for job in jobs:
+            try:
+                jobarr_info = job_info[job.jobid]
+            except KeyError:
+                continue
+
+            # Join the states with ',' in case of job arrays|heterogeneous
+            # jobs
+            job._state = ','.join(m['state'] for m in jobarr_info)
+
+            self._cancel_if_pending_too_long(job)
+            if slurm_state_completed(job.state):
+                # Since Slurm exitcodes are positive take the maximum one
+                job._exitcode = max(
+                    int(m['exit_code'].split(":")[0]) for m in jobarr_info
+                )
+
+            # Use ',' to join nodes to be consistent with Slurm syntax
+            job._nodespec = ','.join(m['nodelist'] for m in jobarr_info)
+
+    def wait(self, job):
+        # Quickly return in case we have finished already
+        self._pull_artefacts(job)
```

Comment:
Pulling the artefacts here could also be problematic and blocking general progress. This is essentially called after the compile or run phase have completed. For all other backends, this returns immediately, whereas here it will block waiting for the artefacts to be pulled back.

Reply:
hm, where would be a better place to do it? When we realize that the job finished in poll? 🤔